datayoga/blocks/expression.py

```diff
@@ -57,24 +92,35 @@ def exec_sql(self, data: Any, expression: str) -> Any:
         Returns:
             Any: Query result
         """
-        clauses = []
-        for k, v in data.items():
-            clauses.append(f"'{v}' as '{k}'")
+        # use a CTE to create the in memory data structure
+        data_inner = data if isinstance(data, list) else [data]
+        cte_clause = self._get_cte(data_inner)
 
-        from_clause = f"select {','.join(clauses)}"
+        column_names = data_inner[0].keys()
 
-        return self.conn.execute(f"select {expression} from ({from_clause})").fetchone()[0]
+        # fetch the CTE and bind the variables
+        data_values = [row.get(colname) for row in data_inner for colname in column_names]
+        self.conn.row_factory = sqlite3.Row
+        cursor = self.conn.execute(
+            f"{cte_clause} select {expression} from data", data_values
+        )
+
+        return cursor.fetchone()[0]
 
 
 class JMESPathExpression(Expression):
     def compile(self, expression: str):
         self.expression = jmespath.compile(expression)
+        self.filter_expression = jmespath.compile(f"[?{expression}]")
+
+    def filter(self, data: Any) -> bool:
```

Comment:
data: Any should be replaced with data: List[Dict[str, Any]]

Reply:
done